### PR TITLE
Feature/basemodel state

### PR DESCRIPTION
**Pull request by MarkDunne**
## Context

It would be nice to move to a Pydantic BaseModel based state for a few reasons

1. Stronger validation of the state. I'm 90% sure there were bugs in the document groups, but we missed this because typeddict wouldn't do any of the validation
2. IDE support
3. We can tack nice things on to the state such as the `last_message` property

## Things to check

- [ ] I have added any new ENV vars in all deployed environments
- [ ] I have tested any code added or changed
- [ ] I have run integration tests

### CODE REVIEW

File: redbox-core/tests/graph/test_state.py [hunk 1]
```diff
@@ -402,7 +410,7 @@ def test_tool_calls_reducer(a: ToolState, b: ToolState, expected: ToolState):
         ),
     ],
 )
-def test_merge_redbox_state_updates(a: RedboxState, b: RedboxState, expected: RedboxState):
+def test_merge_redbox_state_updates(a: dict, b: dict, expected: dict):
```
Comment (MarkDunne) [inline redbox-core/tests/graph/test_state.py:413]: This test never used the documents reducer for some reason, now it does

File: redbox-core/redbox/models/chain.py [hunk 2]
```diff
@@ -261,14 +267,19 @@ def tool_calls_reducer(current: ToolState, update: ToolState | None) -> ToolStat
...
+    route_name: str | None = None
+    tool_calls: Annotated[ToolState | None, tool_calls_reducer] = None
+    metadata: Annotated[RequestMetadata | None, metadata_reducer] = None
+    citations: list[Citation] | None = None
+    steps_left: Annotated[int | None, RemainingStepsManager] = None
+    messages: Annotated[list[AnyMessage], add_messages] = Field(default_factory=list)
+
+    @property
+    def last_message(self) -> AnyMessage:
+        return self.messages[-1]
```
Comment (MarkDunne) [inline redbox-core/redbox/models/chain.py:282]: I think there are probably more nice things we can add like this such as `flat_docs` or `total_token_count`

File: .vscode/redbox.code-workspace [hunk 3]
```diff
@@ -29,4 +29,12 @@
...
 		}
 	],
+	"settings": {
+		"workbench.colorCustomizations": {
+			"activityBar.background": "#911212",
+			"activityBar.foreground": "#fbfbec",
+			"titleBar.activeBackground": "#911212",
+			"titleBar.activeForeground": "#fbfbec",
+		}
+	},
```
Comment (MarkDunne) [inline .vscode/redbox.code-workspace:39]: I use a vscode plugin called [Window Colors](https://marketplace.visualstudio.com/items?itemName=stuart.unique-window-colors) to help me differentiate between projects. This makes redbox red

**Review by jamesrichards4 — APPROVED** [review]

File: redbox-core/redbox/models/chain.py (hunk 2 above)
Comment (gecBurton) [reply redbox-core/redbox/models/chain.py:282]: do we need to catch the edge case where there are no messages?

File: redbox-core/tests/graph/test_state.py (hunk 1 above)
Comment (gecBurton) [reply redbox-core/tests/graph/test_state.py:413]: much clearer :+1:

**Review by gecBurton — APPROVED** [review]
love it ! thank you for preserving with this one
